README.md
# Share your files by FTP 📂

> Implementation of File Transfer Protocol (FTP) based on standard library.


## 🚀 Getting Started

### Install via `go get`
```bash
go get github.com/cxfans/ftplib@v0.1.0
```

### Usage

#### Start a FTP Server
```go
func main() {
	addr := "localhost:2121"
	rootDir := "."
	server, err := ftplib.NewServer(addr, rootDir)
	if err != nil {
		log.Println(err)
	}
	log.Fatal(server.ListenAndServe())
}
```

#### Start a FTP Client
```go
func main() {
	// Login
	c, err := ftplib.Connect("localhost:21", "admin", "admin")
	if err != nil {
		panic(err)
	}
	defer c.Quit()

	// Upload
	data := bytes.NewBufferString("Uploads a file to the remote FTP server.")
	err = c.Stor("file", data)
	if err != nil {
		panic(err)
	}

	// Download
	r, err := c.Retr("file")
	if err != nil {
		panic(err)
	} else {
		buf, err := ioutil.ReadAll(r)
		if err != nil {
			panic(err)
		}
		fp, err := os.Create(file)
		if err != nil {
			panic(err)
		}
		fp.Write(buf)
		r.Close()
	}
}
```

## 🔵 License

This project is licensed under the MIT License - see the [LICENSE](LICENSE.md) file for details.

CHANGELOG.md
# Change Log of ftplib Library

## [0.1.0] - 2019-11-8
### Release
- Implement basic function for File Transfer Protocol (FTP)

utils_test.go
package ftplib

import (
	"fmt"
	"os"
	"testing"
)

// go test -run TestListDetailed
func TestListDetailed(t *testing.T) {
	d, err := os.Open(".")
	if err != nil {
		t.Error(err)
	}
	items, err := d.Readdir(-1)
	if err != nil {
		t.Error(err)
	}
	fmt.Println(string(ListDetailed(items)))
}

// go test -run TestListShort
func TestListShort(t *testing.T) {
	d, err := os.Open(".")
	if err != nil {
		t.Error(err)
	}
	items, err := d.Readdir(-1)
	if err != nil {
		t.Error(err)
	}
	fmt.Println(string(ListShort(items)))
}


client_test.go
package ftplib

import (
	"testing"
)

// go test -run TestConnect
func TestConnect(t *testing.T) {
	c, err := Connect("localhost:2121", "up", "up")
	if err != nil {
		t.Error(err)
	}
	_ = c.Quit()
}

func TestConnectAnonymous(t *testing.T) {
	c, err := ConnectAnonymous("localhost:2121")
	if err != nil {
		t.Error(err)
	}
	_ = c.Quit()
}


client.go
package ftplib

import (
	"bufio"
	"errors"
	"io"
	"log"
	"net"
	"net/textproto"
	"strconv"
	"strings"
	"time"
)

/// EntryType describes the different types of an Entry.
type EntryType int

const (
	EntryTypeFile EntryType = iota
	EntryTypeFolder
	EntryTypeLink
)

// ClientConn represents the connection to a remote FTP server.
type ClientConn struct {
	conn     *textproto.Conn
	host     string
	timeout  time.Duration
	features map[string]string
}

// response represent a data-connection
type response struct {
	conn net.Conn
	c    *ClientConn
}

// ClientConn represents the connection to a remote FTP server.
type Entry struct {
	Name string
	Type EntryType
	Size uint64
	Time time.Time
}

func (c *ClientConn) Quit() error {
	c.conn.Cmd("QUIT")
	return c.conn.Close()
}

func (r *response) Close() error {
	err := r.conn.Close()
	_, _, err2 := r.c.conn.ReadResponse(StatusClosingDataConnection)
	if err2 != nil {
		err = err2
	}
	return err
}

func (r *response) Read(buf []byte) (int, error) {
	return r.conn.Read(buf)
}

// Dial is like DialTimeout with no timeout
func Dial(addr string) (*ClientConn, error) {
	return DialTimeout(addr, 0)
}

func DialTimeout(addr string, timeout time.Duration) (*ClientConn, error) {
	tconn, err := net.DialTimeout("tcp", addr, timeout)
	if err != nil {
		return nil, err
	}
	// Use the resolved IP address in case addr contains a domain name
	// If we use the domain name, we might not resolve to the same IP.
	remoteAddr := tconn.RemoteAddr().(*net.TCPAddr)
	conn := textproto.NewConn(tconn)

	c := &ClientConn{
		conn:     conn,
		host:     remoteAddr.IP.String(),
		timeout:  timeout,
		features: make(map[string]string),
	}

	_, msg, err := c.conn.ReadResponse(StatusReady)
	//_, msg, err := c.conn.ReadResponse(StatusReady)
	log.Println(msg)
	if err != nil {
		c.Quit()
		return nil, err
	}

	err = c.feat()
	if err != nil {
		c.Quit()
		return nil, err
	}

	err = c.setUTF8()
	if err != nil {
		c.Quit()
		return nil, err
	}

	return c, nil
}

// setUTF8 issues an "OPTS UTF8 ON" command.
func (c *ClientConn) setUTF8() error {
	if _, ok := c.features["UTF8"]; !ok {
		return nil
	}

	code, message, err := c.cmd(-1, "OPTS UTF8 ON")
	if err != nil {
		return err
	}

	if code != StatusCommandOK {
		return errors.New(message)
	}

	log.Println("Set utf-8")

	return nil
}

// feat issues a FEAT FTP command to list the additional commands supported by
// the remote FTP server.
// FEAT is described in RFC 2389
func (c *ClientConn) feat() error {
	code, message, err := c.cmd(-1, "FEAT")
	if err != nil {
		return err
	}

	if code != StatusSystem {
		// The server does not support the FEAT command. This is not an
		// error: we consider that there is no additional feature.
		return nil
	}

	lines := strings.Split(message, "\n")
	for _, line := range lines {
		if !strings.HasPrefix(line, " ") {
			continue
		}

		line = strings.TrimSpace(line)
		featureElements := strings.SplitN(line, " ", 2)

		command := featureElements[0]

		var commandDesc string
		if len(featureElements) == 2 {
			commandDesc = featureElements[1]
		}

		c.features[command] = commandDesc
	}

	return nil
}

// Login authenticates the client with specified user and password.
//
// "anonymous"/"anonymous" is a common user/password scheme for FTP servers
// that allows anonymous read-only accounts.
func (c *ClientConn) Login(user, password string) error {
	code, message, err := c.cmd(-1, "USER %s", user)
	if err != nil {
		return err
	}

	switch code {
	case StatusLoggedIn:
	case StatusUserOK:
		_, _, err = c.cmd(StatusLoggedIn, "PASS %s", password)
		if err != nil {
			return err
		}
	default:
		return errors.New(message)
	}

	// Switch to binary mode
	_, _, err = c.cmd(StatusCommandOK, "TYPE I")
	if err != nil {
		return err
	}

	log.Println("User logged in.")
	return nil
}

// Logout issues a REIN FTP command to logout the current user.
func (c *ClientConn) Logout() error {
	_, _, err := c.cmd(StatusReady, "REIN")
	return err
}

func ConnectAnonymous(addr string) (*ClientConn, error) {
	return Connect(addr, "anonymous", "anonymous")
}

func Connect(addr, user, password string) (*ClientConn, error) {
	c, err := Dial(addr)
	if err != nil {
		c.Quit()
		return nil, err
	}
	return c, c.Login(user, password)
}

// epsv issues an "EPSV" command to get a port number for a data connection.
// Enter extended passive mode
func (c *ClientConn) epsv() (port int, err error) {
	_, line, err := c.cmd(StatusExtendedPassiveMode, "EPSV")
	if err != nil {
		return
	}
	// 229 Entering Extended Passive Mode (|||65202|)
	start := strings.Index(line, "|||")
	end := strings.LastIndex(line, "|")
	if start == -1 || end == -1 {
		err = errors.New("invalid EPSV response format")
		return
	}
	port, err = strconv.Atoi(line[start+3 : end])
	return
}

// Enter passive mode
func (c *ClientConn) pasv() (port int, err error) {
	_, line, err := c.cmd(StatusPassiveMode, "PASV")
	if err != nil {
		return
	}
	// 227 Entering Passive Mode (172,17,66,241,254,179)
	start := strings.Index(line, "(")
	end := strings.LastIndex(line, ")")
	if start == -1 || end == -1 {
		err = errors.New("invalid PASV response format")
		return
	}
	lst := strings.Split(string(line[start+1:end]), ",")
	n := len(lst)
	x, _ := strconv.Atoi(lst[n-2])
	y, _ := strconv.Atoi(lst[n-1])
	port, err = x*256+y, nil
	return
}

// openDataConn creates a new FTP data connection.
func (c *ClientConn) openDataConn() (net.Conn, error) {
	var (
		port int
		err  error
	)

	if port, err = c.epsv(); err != nil {
		if port, err = c.pasv(); err != nil {
		}
	}

	return net.DialTimeout("tcp", net.JoinHostPort(c.host, strconv.Itoa(port)), c.timeout)
}

// parseListLine parses the various non-standard
// format returned by the LIST FTP command.
func parseListLine(line string) (*Entry, error) {
	fields := strings.Fields(line)
	if len(fields) < 9 {
		return nil, errors.New("unsupported LIST line")
	}

	e := &Entry{}
	switch fields[0][0] {
	case '-':
		e.Type = EntryTypeFile
	case 'd':
		e.Type = EntryTypeFolder
	case 'l':
		e.Type = EntryTypeLink
	default:
		return nil, errors.New("unknown entry type")
	}

	if e.Type == EntryTypeFile {
		size, err := strconv.ParseUint(fields[4], 10, 0)
		if err != nil {
			return nil, err
		}
		e.Size = size
	}
	var timeStr string
	if strings.Contains(fields[7], ":") { // this year
		thisYear, _, _ := time.Now().Date()
		timeStr = fields[6] + " " + fields[5] + " " +
			strconv.Itoa(thisYear)[2:4] + " " + fields[7] + " GMT"
	} else { // not this year
		timeStr = fields[6] + " " + fields[5] + " " + fields[7][2:4] + " " + "00:00" + " GMT"
	}
	t, err := time.Parse("_2 Jan 06 15:04 MST", timeStr)
	if err != nil {
		return nil, err
	}
	e.Time = t

	e.Name = strings.Join(fields[8:], " ")
	return e, nil
}

// NameList issues an NLST FTP command.
func (c *ClientConn) NameList(path string) (entries []string, err error) {
	conn, err := c.cmdDataConnFrom(0, "NLST %s", path)
	if err != nil {
		return
	}

	r := &response{conn, c}
	defer r.Close()

	scanner := bufio.NewScanner(r)
	for scanner.Scan() {
		entries = append(entries, scanner.Text())
	}
	if err = scanner.Err(); err != nil {
		return entries, err
	}
	return
}

// List issues a LIST FTP command.
func (c *ClientConn) List(path string) (entries []*Entry, err error) {
	conn, err := c.cmdDataConnFrom(0, "LIST %s", path)
	if err != nil {
		return
	}
	r := &response{conn, c}
	defer r.Close()

	bio := bufio.NewReader(r)
	for {
		line, e := bio.ReadString('\n')
		if e == io.EOF {
			break
		} else if e != nil {
			return nil, e
		}
		entry, err := parseListLine(line)
		if err == nil {
			entries = append(entries, entry)
		}
	}
	return
}

func (c *ClientConn) ChangeDir(path string) error {
	_, _, err := c.cmd(StatusRequestedFileActionOK, "CWD %s", path)
	return err
}

// Changes the current directory to the parent directory.
// ChangeDir("..")
func (c *ClientConn) ChangeDirToParent() error {
	_, _, err := c.cmd(StatusRequestedFileActionOK, "CDUP")
	return err
}

// Returns the path of the current directory.
func (c *ClientConn) CurrentDir() (string, error) {
	_, msg, err := c.cmd(StatusPathCreated, "PWD")
	if err != nil {
		return "", err
	}

	start := strings.Index(msg, "\"")
	end := strings.LastIndex(msg, "\"")

	if start == -1 || end == -1 {
		return "", errors.New("unsupported PWD response format")
	}

	return msg[start+1 : end], nil
}

// Retrieves a file from the remote FTP server.
// The ReadCloser must be closed at the end of the operation.
func (c *ClientConn) Retr(path string) (io.ReadCloser, error) {
	return c.RetrFrom(path, 0)
}

// Retr issues a RETR FTP command to fetch the specified file from the remote
// FTP server, the server will not send the offset first bytes of the file.
func (c *ClientConn) RetrFrom(path string, offset uint64) (io.ReadCloser, error) {
	conn, err := c.cmdDataConnFrom(offset, "RETR %s", path)
	if err != nil {
		return nil, err
	}

	return &response{conn, c}, nil
}

// Uploads a file to the remote FTP server.
// This function gets the data from the io.Reader. Hint: io.Pipe()
func (c *ClientConn) Stor(path string, r io.Reader) error {
	return c.StorFrom(path, r, 0)
}

// Stor issues a STOR FTP command to store a file to the remote FTP server.
// Stor creates the specified file with the content of the io.Reader, writing
// on the server will start at the given file offset.
func (c *ClientConn) StorFrom(path string, r io.Reader, offset uint64) error {
	conn, err := c.cmdDataConnFrom(offset, "STOR %s", path)

	if err != nil {
		return err
	}

	_, err = io.Copy(conn, r)
	conn.Close()
	if err != nil {
		return err
	}

	_, _, err = c.conn.ReadResponse(StatusClosingDataConnection)
	return err
}

func (c *ClientConn) Rename(from, to string) error {
	_, _, err := c.cmd(StatusRequestFilePending, "RNFR %s", from)
	if err != nil {
		return err
	}

	_, _, err = c.cmd(StatusRequestedFileActionOK, "RNTO %s", to)
	return err
}

// Creates a new directory on the remote FTP server.
func (c *ClientConn) MakeDir(path string) error {
	_, _, err := c.cmd(StatusPathCreated, "MKD %s", path)
	return err
}

// Removes a directory from the remote FTP server.
func (c *ClientConn) RemoveDir(path string) error {
	_, _, err := c.cmd(StatusRequestedFileActionOK, "RMD %s", path)
	return err
}

// Deletes a file on the remote FTP server.
func (c *ClientConn) Delete(path string) error {
	_, _, err := c.cmd(StatusRequestedFileActionOK, "DELE %s", path)
	return err
}

// Sends a NOOP command. Usualy used to prevent timeouts.
func (c *ClientConn) NoOp() error {
	_, _, err := c.cmd(StatusCommandOK, "NOOP")
	return err
}

// cmd is a helper function to execute a command and
// check for the expected FTP return code
func (c *ClientConn) cmd(expected int, format string, args ...interface{}) (int, string, error) {
	_, err := c.conn.Cmd(format, args...)
	if err != nil {
		return 0, "", err
	}

	return c.conn.ReadResponse(expected)
}

// cmdDataConnFrom executes a command which require a FTP data connection.
// Issues a REST FTP command to specify the number of bytes to skip for the transfer.
func (c *ClientConn) cmdDataConnFrom(offset uint64, format string, args ...interface{}) (net.Conn, error) {
	conn, err := c.openDataConn()
	if err != nil {
		return nil, err
	}

	if offset != 0 {
		_, _, err := c.cmd(StatusRequestFilePending, "REST %d", offset)
		if err != nil {
			conn.Close()
			return nil, err
		}
	}

	_, err = c.conn.Cmd(format, args...)
	if err != nil {
		conn.Close()
		return nil, err
	}
	code, msg, err := c.conn.ReadResponse(-1)
	if err != nil {
		conn.Close()
		return nil, err
	}
	if code != StatusAlreadyOpen && code != StatusAboutToSend {
		conn.Close()
		// It easier for the client to extract the code and message with type assertions.
		return nil, &textproto.Error{code, msg}
	}
	return conn, nil
}


status.go
package ftplib

// Status codes, defined in RFC 959
const (
	StatusInitiating    = 100
	StatusRestartMarker = 110
	StatusReadyMinute   = 120
	StatusAlreadyOpen   = 125
	StatusAboutToSend   = 150

	StatusCommandOK             = 200
	StatusCommandNotImplemented = 202
	StatusSystem                = 211
	StatusDirectory             = 212
	StatusFile                  = 213
	StatusHelp                  = 214
	StatusName                  = 215
	StatusReady                 = 220
	StatusClosing               = 221
	StatusDataConnectionOpen    = 225
	StatusClosingDataConnection = 226
	StatusPassiveMode           = 227
	StatusLongPassiveMode       = 228
	StatusExtendedPassiveMode   = 229
	StatusLoggedIn              = 230
	StatusLoggedOut             = 231
	StatusLogoutAck             = 232
	StatusRequestedFileActionOK = 250
	StatusPathCreated           = 257

	StatusUserOK             = 331
	StatusLoginNeedAccount   = 332
	StatusRequestFilePending = 350

	StatusNotAvailable             = 421
	StatusCanNotOpenDataConnection = 425
	StatusTransfertAborted         = 426
	StatusInvalidCredentials       = 430
	StatusHostUnavailable          = 434
	StatusFileActionIgnored        = 450
	StatusActionAborted            = 451
	Status452                      = 452

	StatusBadCommand              = 500
	StatusBadArguments            = 501
	StatusNotImplemented          = 502
	StatusBadSequence             = 503
	StatusNotImplementedParameter = 504
	StatusNotLoggedIn             = 530
	StatusStorNeedAccount         = 532
	StatusFileUnavailable         = 550
	StatusPageTypeUnknown         = 551
	StatusExceededStorage         = 552
	StatusBadFileName             = 553
)

var messages = map[int]string{
	// 200
	StatusCommandOK:             "Command okay.",
	StatusCommandNotImplemented: "Command not implemented, superfluous at this site.",
	StatusSystem:                "System status, or system help reply.",
	StatusDirectory:             "Directory status.",
	StatusFile:                  "File status.",
	StatusHelp:                  "Help message.",
	StatusName:                  "UNIX Type: L8",
	StatusReady:                 "Service ready for new user.",
	StatusClosing:               "Service closing control connection.",
	StatusDataConnectionOpen:    "Data connection open; no transfer in progress.",
	StatusClosingDataConnection: "Closing data connection. Requested file action successful.",
	StatusPassiveMode:           "Entering Passive Mode.",
	StatusLongPassiveMode:       "Entering Long Passive Mode.",
	StatusExtendedPassiveMode:   "Entering Extended Passive Mode.",
	StatusLoggedIn:              "User logged in, proceed.",
	StatusLoggedOut:             "User logged out; service terminated.",
	StatusLogoutAck:             "Logout command noted, will complete when transfer done.",
	StatusRequestedFileActionOK: "Requested file action okay, completed.",
	StatusPathCreated:           "Path created.",

	// 300
	StatusUserOK:             "User name okay, need password.",
	StatusLoginNeedAccount:   "Need account for login.",
	StatusRequestFilePending: "Requested file action pending further information.",

	// 400
	StatusNotAvailable:             "Service not available, closing control connection.",
	StatusCanNotOpenDataConnection: "Can't open data connection.",
	StatusTransfertAborted:         "Connection closed; transfer aborted.",
	StatusInvalidCredentials:       "Invalid username or password.",
	StatusHostUnavailable:          "Requested host unavailable.",
	StatusFileActionIgnored:        "Requested file action not taken.",
	StatusActionAborted:            "Requested action aborted. Local error in processing.",
	Status452:                      "Insufficient storage space in system.",

	// 500
	StatusBadCommand:              "Command unrecognized.",
	StatusBadArguments:            "Syntax error in parameters or arguments.",
	StatusNotImplemented:          "Command not implemented.",
	StatusBadSequence:             "Bad sequence of commands.",
	StatusNotImplementedParameter: "Command not implemented for that parameter.",
	StatusNotLoggedIn:             "Not logged in.",
	StatusStorNeedAccount:         "Need account for storing files.",
	StatusFileUnavailable:         "File unavailable.",
	StatusPageTypeUnknown:         "Page type unknown.",
	StatusExceededStorage:         "Exceeded storage allocation.",
	StatusBadFileName:             "File name not allowed.",
}

// Returns a message for different status codes
func Message(code int) string {
	return messages[code]
}


commands.go
package ftplib

// Commands, defined in RFC 959
const (
	ABOR = "ABOR" // Abort an active file transfer.
	ACCT = "ACCT" // Account information.
	ADAT = "ADAT" // Authentication/Security Data
	ALLO = "ALLO" // Allocate sufficient disk space to receive a file.
	APPE = "APPE" // Append (with create)
	AUTH = "AUTH" // Authentication/Security Mechanism
	AVBL = "AVBL" // Get the available space
	CCC  = "CCC"  // Clear Command Channel
	CDUP = "CDUP" // Change to Parent Directory.
	CONF = "CONF" // Confidentiality Protection Command
	CSID = "CSID" // Client / Server Identification
	CWD  = "CWD"  // Change working directory.
	DELE = "DELE" // Delete file.
	DSIZ = "DSIZ" // Get the directory size
	ENC  = "ENC"  // Privacy Protected Channel
	EPRT = "EPRT" // Specifies an extended address and port to which the server should connect.
	EPSV = "EPSV" // Enter extended passive mode.
	FEAT = "FEAT" // Get the feature list implemented by the server.
	HELP = "HELP" // Returns usage documentation on a command if specified, else a general help document is returned.
	HOST = "HOST" // Identify desired virtual host on server, by name.
	LANG = "LANG" // Language Negotiation
	LIST = "LIST" // Returns information of a file or directory if specified, else information of the current working directory is returned.
	LPRT = "LPRT" // Specifies a long address and port to which the server should connect.
	LPSV = "LPSV" // Enter long passive mode.
	MDTM = "MDTM" // Return the last-modified time of a specified file.
	MFCT = "MFCT" // Modify the creation time of a file.
	MFF  = "MFF"  // Modify fact (the last modification time, creation time, UNIX group/owner/mode of a file).
	MFMT = "MFMT" // Modify the last modification time of a file.
	MIC  = "MIC"  // Integrity Protected Command
	MKD  = "MKD"  // Make directory.
	MLSD = "MLSD" // Lists the contents of a directory if a directory is named.
	MLST = "MLST" // Provides data about exactly the object named on its command line, and no others.
	MODE = "MODE" // Sets the transfer mode (Stream, Block, or Compressed).
	NLST = "NLST" // Returns a list of file names in a specified directory.
	NOOP = "NOOP" // No operation (dummy packet; used mostly on keepalives).
	OPTS = "OPTS" // Select options for a feature (for example OPTS UTF8 ON).
	PASS = "PASS" // Authentication password.
	PASV = "PASV" // Enter passive mode.
	PBSZ = "PBSZ" // Protection Buffer Size
	PORT = "PORT" // Specifies an address and port to which the server should connect.
	PROT = "PROT" // Data Channel Protection Level.
	PWD  = "PWD"  // Print working directory. Returns the current directory of the host.
	QUIT = "QUIT" // Disconnect.
	REIN = "REIN" // Re initializes the connection.
	REST = "REST" // Restart transfer from the specified point.
	RETR = "RETR" // Retrieve a copy of the file
	RMD  = "RMD"  // Remove a directory.
	RMDA = "RMDA" // Remove a directory tree
	RNFR = "RNFR" // Rename from.
	RNTO = "RNTO" // Rename to.
	SITE = "SITE" // Sends site specific commands to remote server (like SITE IDLE 60or SITE UMASK 002). Inspect SITE HELP output for complete list of supported commands.
	SIZE = "SIZE" // Return the size of a file.
	SMNT = "SMNT" // Mount file structure.
	SPSV = "SPSV" // Use single port passive mode (only one TCP port number for both control connections and passive-mode data connections)
	STAT = "STAT" // Returns the current status.
	STOR = "STOR" // Accept the data and to store the data as a file at the server site
	STOU = "STOU" // Store file uniquely.
	STRU = "STRU" // Set file transfer structure.
	SYST = "SYST" // Return system type.
	THMB = "THMB" // Get a thumbnail of a remote image file
	TYPE = "TYPE" // Sets the transfer mode (ASCII/Binary).
	USER = "USER" // Authentication username.
	XCUP = "XCUP" // Change to the parent of the current working directory
	XMKD = "XMKD" // Make a directory
	XPWD = "XPWD" // Print the current working directory
	XRCP = "XRCP" //
	XRMD = "XRMD" // Remove the directory
	XRSQ = "XRSQ" //
	XSEM = "XSEM" // Send, mail if cannot
	XSEN = "XSEN" // Send to terminal
)


dataconn.go
package ftplib

import (
	"log"
	"net"
	"strconv"
	"strings"
)

type DataConn interface {
	Host() string
	Port() int
	Write(p []byte) (n int, err error)
	Read(p []byte) (n int, err error)
	Close() error
}

type PassiveConn struct {
	conn       *net.TCPConn
	host, port string
	done       chan bool
	err        error
}

func NewPassiveConn(host string) (passiveConn *PassiveConn, err error) {
	passiveConn = &PassiveConn{host: host, done: make(chan bool, 1)}
	if err := passiveConn.ListenAndServe(); err != nil {
		return nil, err
	}
	log.Println("A new passive connection created.")
	return passiveConn, nil
}

func (passiveConn *PassiveConn) Host() string {
	return "127.0.0.1"
}

func (passiveConn *PassiveConn) Port() int {
	port, _ := strconv.Atoi(passiveConn.port)
	return port
}

func (passiveConn *PassiveConn) Close() error {
	log.Println("Passive data connection closed.")
	return passiveConn.conn.Close()
}

func (passiveConn *PassiveConn) ListenAndServe() error {
	laddr, err := net.ResolveTCPAddr("tcp4", passiveConn.host+":0")
	if err != nil {
		log.Println(err)
		return err
	}
	listener, err := net.ListenTCP("tcp4", laddr)
	if err != nil {
		log.Println(err)
		return err
	}
	addr := listener.Addr()
	parts := strings.Split(addr.String(), ":")
	passiveConn.host = parts[0]
	passiveConn.port = parts[1]

	go func() {
		conn, err := listener.AcceptTCP()
		passiveConn.done <- true
		if err != nil {
			log.Println(err)
			passiveConn.err = err
			return
		}
		passiveConn.err = nil
		passiveConn.conn = conn
	}()

	return nil
}

func (passiveConn *PassiveConn) wait() bool {
	if passiveConn.conn != nil {
		return true
	}
	<-passiveConn.done
	return passiveConn.conn != nil
}

func (passiveConn *PassiveConn) Read(data []byte) (n int, err error) {
	if !passiveConn.wait() {
		return 0, passiveConn.err
	}
	return passiveConn.conn.Read(data)
}

func (passiveConn *PassiveConn) Write(data []byte) (n int, err error) {
	if !passiveConn.wait() {
		return 0, passiveConn.err
	}
	return passiveConn.conn.Write(data)
}


server.go
/*
	Implements file transfer protocol server based on standard library
*/

package ftplib

import (
	"bufio"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net"
	"os"
	"path"
	"strconv"
	"strings"
)

type Server struct {
	listener *net.TCPListener
	host     string
	rootDir  string
}

func NewServer(addr, rootDir string) (server *Server, err error) {
	laddr, err := net.ResolveTCPAddr("tcp4", addr)
	if err != nil {
		log.Println(err)
		return nil, err
	}
	listener, err := net.ListenTCP("tcp4", laddr)
	if err != nil {
		log.Println(err)
		return nil, err
	}
	host, _, _ := net.SplitHostPort(addr)
	return &Server{listener: listener, host: host, rootDir: rootDir}, nil
}

func (server *Server) ListenAndServe() (err error) {
	log.Println("Server start.")
	for {
		conn, err := server.listener.AcceptTCP()
		if err != nil {
			log.Println(err)
			return err
		}

		serverConn := &ServerConn{
			conn:   conn,
			reader: bufio.NewReader(conn),
			writer: bufio.NewWriter(conn),
			prefix: server.rootDir,
			host:   server.host,
		}

		log.Println(conn.RemoteAddr().String(), "connected.")

		go serverConn.Serve()
	}
}

func (server *Server) Stop() (err error) {
	if server.listener != nil {
		return server.listener.Close()
	}
	return nil
}

type ServerConn struct {
	conn             *net.TCPConn
	reader           *bufio.Reader
	writer           *bufio.Writer
	dataConn         DataConn
	prefix, host, rn string
}

func (serverConn *ServerConn) Close() {
	serverConn.conn.Close()
	if serverConn.dataConn != nil {
		serverConn.dataConn.Close()
		serverConn.dataConn = nil
	}
	log.Println("Closed one connection.")
}

func (serverConn *ServerConn) cmd(msg string, v ...interface{}) (n int) {
	n, err := serverConn.writer.WriteString(msg + "\r\n")
	if err != nil {
		log.Println(err)
		serverConn.Close()
	}
	serverConn.writer.Flush()
	log.Println(msg)
	return
}

func (serverConn *ServerConn) sendCodeLine(code int, msg string) {
	serverConn.cmd(fmt.Sprintf("%d %s", code, msg))
}

func (serverConn *ServerConn) sendStatusText(code int) {
	serverConn.sendCodeLine(code, Message(code))
}

func (serverConn *ServerConn) sendData(data []byte) {
	if serverConn.dataConn != nil {
		n, _ := serverConn.dataConn.Write(data)
		serverConn.dataConn.Close()
		msg := fmt.Sprintf("Closing data connection, sent %d bytes.", n)
		serverConn.sendCodeLine(StatusClosingDataConnection, msg)
	} else {
		serverConn.sendStatusText(StatusTransfertAborted)
	}
}

func (serverConn *ServerConn) parsingPath(params []string) string {
	p := strings.Join(params, " ")
	if strings.HasPrefix(p, "/") {
		p = path.Join(".", p)
	} else {
		p = path.Join(serverConn.prefix, p)
	}
	return p
}

func (serverConn *ServerConn) Serve() {
	log.Println("Connection established: start server.")
	serverConn.sendStatusText(StatusReady)

loop:
	for {
		cmdLine, err := serverConn.reader.ReadString('\n')
		log.Print(cmdLine)
		if err != nil {
			// When the client closes the connection, the server will read EOF.
			if err == io.EOF {
				break loop
			}
			log.Println("Loop Error:", err)
			serverConn.Close()
			break loop
		}
		params := strings.Split(strings.TrimSpace(cmdLine), " ")
		switch strings.ToUpper(params[0]) {

		case USER:
			serverConn.sendStatusText(StatusUserOK)

		case PASS:
			serverConn.sendStatusText(StatusLoggedIn)

		case PWD:
			serverConn.sendCodeLine(StatusPathCreated,
				fmt.Sprintf("\"%s\" is current directory.", serverConn.prefix))

		case CWD:
			p := serverConn.parsingPath(params[1:])
			f, err := os.Stat(p)
			if f.IsDir() && err == nil {
				serverConn.prefix = p
				serverConn.sendCodeLine(StatusRequestedFileActionOK,
					"Directory changed to "+serverConn.prefix)
			} else {
				serverConn.sendStatusText(StatusFileUnavailable)
			}

		case DELE:
			p := serverConn.parsingPath(params[1:])
			_, err := os.Stat(p)
			if err != nil {
				serverConn.sendStatusText(StatusFileUnavailable)
			} else {
				os.Remove(p)
				serverConn.sendCodeLine(StatusRequestedFileActionOK, "File deleted.")
			}

		case EPSV:
			passiveConn, err := NewPassiveConn(serverConn.host)
			if err != nil {
				serverConn.sendStatusText(StatusCanNotOpenDataConnection)
			} else {
				serverConn.dataConn = passiveConn
				serverConn.sendCodeLine(StatusExtendedPassiveMode,
					fmt.Sprintf("Entering Extended Passive Mode (|||%d|)", passiveConn.Port()))
			}

		case SIZE:
			p := serverConn.parsingPath(params[1:])
			f, err := os.Stat(p)
			if f.IsDir() && err == nil {
				serverConn.sendCodeLine(StatusFile, "1024")
			} else {
				serverConn.sendCodeLine(StatusFile, strconv.Itoa(int(f.Size())))
			}

		case LIST:
			serverConn.sendCodeLine(StatusAboutToSend,
				"Opening ASCII mode data connection for file list")
			d, _ := os.Open(serverConn.prefix)
			items, _ := d.Readdir(-1)
			info := ListDetailed(items)
			serverConn.sendData(info)

		case MKD:
			p := serverConn.parsingPath(params[1:])
			err = os.Mkdir(p, 0777)
			if err == nil {
				serverConn.sendStatusText(StatusPathCreated)
			} else {
				serverConn.sendCodeLine(StatusFileUnavailable, fmt.Sprint(err))
			}

		case NOOP:
			serverConn.sendStatusText(StatusCommandOK)

		case PASV:
			passiveConn, err := NewPassiveConn(serverConn.host)
			if err != nil {
				serverConn.sendStatusText(StatusCanNotOpenDataConnection)
			} else {
				serverConn.dataConn = passiveConn
				port := passiveConn.Port()
				x := port / 256
				y := port - x*256
				quad := strings.ReplaceAll(passiveConn.Host(), ".", ",")
				msg := fmt.Sprintf("Entering Passive Mode (%s,%d,%d)", quad, x, y)
				serverConn.sendCodeLine(227, msg)
			}

		case QUIT:
			serverConn.Close()
			break loop

		case RETR:
			p := serverConn.parsingPath(params[1:])
			data, err := ioutil.ReadFile(p)
			if err != nil {
				serverConn.sendCodeLine(StatusFileUnavailable, fmt.Sprint(err))
			} else {
				bytes := strconv.Itoa(len(data))
				serverConn.sendCodeLine(150, "Data transfer starting "+bytes+"bytes")
				serverConn.sendData([]byte(data))
			}

		case RMD, XRMD:
			p := serverConn.parsingPath(params[1:])
			f, err := os.Stat(p)
			if f.IsDir() && err == nil {
				err := os.RemoveAll(p)
				if err != nil {
					serverConn.sendCodeLine(StatusFileUnavailable, fmt.Sprint(err))
				} else {
					serverConn.sendCodeLine(StatusRequestedFileActionOK, "Directory deleted.")
				}
			} else {
				serverConn.sendStatusText(StatusFileUnavailable)
			}

		case RNFR:
			serverConn.rn = serverConn.parsingPath(params[1:])
			serverConn.sendStatusText(StatusRequestFilePending)

		case RNTO:
			p := serverConn.parsingPath(params[1:])
			err := os.Rename(serverConn.rn, p)
			if err != nil {
				serverConn.sendCodeLine(StatusFileUnavailable, fmt.Sprint(err))
			} else {
				serverConn.sendCodeLine(StatusRequestedFileActionOK, "File renamed.")
			}

		case SYST:
			serverConn.sendStatusText(StatusName)

		case STOR:
			p := serverConn.parsingPath(params[1:])
			serverConn.sendCodeLine(150, "Data transfer starting.")
			file, err := os.OpenFile(p, os.O_WRONLY|os.O_TRUNC|os.O_CREATE,
				0666)
			if err != nil {
				serverConn.sendStatusText(450)
			}
			n, _ := io.Copy(file, serverConn.dataConn)

			if n >= 0 {
				serverConn.sendCodeLine(226, "OK, received "+
					strconv.Itoa(int(n))+" bytes.")
			} else {
				serverConn.sendStatusText(550)
			}
			file.Close()

		case TYPE:
			param := strings.ToUpper(params[1])
			if param == "A" {
				serverConn.sendCodeLine(StatusCommandOK, "Type set to ASCII.")
			} else if param == "I" {
				serverConn.sendCodeLine(StatusCommandOK, "Type set to binary.")
			} else {
				serverConn.sendCodeLine(StatusBadArguments, "Invalid type.")
			}

		default:
			serverConn.sendStatusText(StatusCommandNotImplemented)

		}
	}
	log.Println("Connection terminated: stop server.")
}
